prov/efa/src/efa_domain.c

```diff
@@ -430,3 +437,203 @@ efa_domain_ops_open(struct fid *fid, const char *ops_name, uint64_t flags,
 
 	return ret;
 }
+
+
+void efa_domain_progress_rdm_peers_and_queues(struct efa_domain *domain)
+{
+	struct efa_rdm_peer *peer;
+	struct dlist_entry *tmp;
+	struct efa_rdm_ope *ope;
+	int ret;
+
+	assert(domain->info->ep_attr->type == FI_EP_RDM);
+
+	/* Update timers for peers that are in backoff list*/
+	dlist_foreach_container_safe(&domain->peer_backoff_list, struct efa_rdm_peer,
+				     peer, rnr_backoff_entry, tmp) {
+		if (ofi_gettime_us() >= peer->rnr_backoff_begin_ts +
+					peer->rnr_backoff_wait_time) {
+			peer->flags &= ~EFA_RDM_PEER_IN_BACKOFF;
+			dlist_remove(&peer->rnr_backoff_entry);
+		}
+	}
+
+	/*
+	 * Resend handshake packet for any peers where the first
+	 * handshake send failed.
+	 */
+	dlist_foreach_container_safe(&domain->handshake_queued_peer_list,
+				     struct efa_rdm_peer, peer,
+				     handshake_queued_entry, tmp) {
+		if (peer->flags & EFA_RDM_PEER_IN_BACKOFF)
+			continue;
+
+		ret = efa_rdm_ep_post_handshake(peer->ep, peer);
+		if (ret == -FI_EAGAIN)
+			break;
+
+		if (OFI_UNLIKELY(ret)) {
+			EFA_WARN(FI_LOG_EP_CTRL,
+				"Failed to post HANDSHAKE to peer %ld: %s\n",
+				peer->efa_fiaddr, fi_strerror(-ret));
+			efa_base_ep_write_eq_error(&peer->ep->base_ep, -ret, FI_EFA_ERR_PEER_HANDSHAKE);
+			return;
+		}
+
+		dlist_remove(&peer->handshake_queued_entry);
+		peer->flags &= ~EFA_RDM_PEER_HANDSHAKE_QUEUED;
+		peer->flags |= EFA_RDM_PEER_HANDSHAKE_SENT;
+	}
+
+	/*
+	 * Resend queued RNR pkts
+	 */
+	dlist_foreach_container_safe(&domain->ope_queued_rnr_list,
+				     struct efa_rdm_ope,
+				     ope, queued_rnr_entry, tmp) {
+		peer = efa_rdm_ep_get_peer(ope->ep, ope->addr);
+		assert(peer);
+
+		if (peer->flags & EFA_RDM_PEER_IN_BACKOFF)
+			continue;
+
+		assert(ope->internal_flags & EFA_RDM_OPE_QUEUED_RNR);
+		assert(!dlist_empty(&ope->queued_pkts));
+		ret = efa_rdm_ep_post_queued_pkts(ope->ep, &ope->queued_pkts);
+
+		if (ret == -FI_EAGAIN)
+			break;
+
+		if (OFI_UNLIKELY(ret)) {
+			assert(ope->type == EFA_RDM_RXE || ope->type == EFA_RDM_TXE);
+			if (ope->type == EFA_RDM_RXE)
+				efa_rdm_rxe_handle_error(ope, -ret, FI_EFA_ERR_PKT_SEND);
+			else
+				efa_rdm_txe_handle_error(ope, -ret, FI_EFA_ERR_PKT_SEND);
```

Comment:
Future improvement: At some point we should consider adding function pointers to `struct efa_rdm_ope` (or a dispatch table) for functions like `efa_rdm_[rt]xe_handle_error`, `efa_rdm_[rt]xe_release`, etc to alleviate the need to check the `type` before dispatching to functions where we're passing the same arguments anyway.